src/packet/gre.rs.in

```diff
@@ -0,0 +1,125 @@
+// Copyright (c) 2016 Robert Collins <robertc@robertcollins.net>
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+#[cfg(test)]
+use packet::Packet;
+
+use packet::PrimitiveValues;
+
+use pnet_macros_support::types::*;
+
+/// GRE Packet.
+///
+/// Note that routing information from RFC 1701 is not implemented, packets
+/// with routing_present true will have incorrect payload lengths.
+#[packet]
+pub struct Gre {
+    checksum_present: u1,
+    routing_present: u1,
+    key_present: u1,
+    sequence_present: u1,
+    strict_source_route: u1,
+    recursion_control: u3,
+    zero_flags: u5,
+    version: u3,
+    protocol_type: u16be, // 0x800 for ipv4 [basically an ethertype
+    #[length_fn = "gre_checksum_length"]
+    checksum: Vec<U16BE>,
```

Comment:
This is very unfortunate, and seems to be occurring more often. Perhaps it would be worth adding support for `Option<T>` to pnet_macros.